hathor/indexes/rocksdb_deps_index.py

```diff
@@ -0,0 +1,413 @@
+            tx_hash = key[1:]
+            assert len(tx_hash) == 32
+            return _KeyAny(tag, tx_hash, None)
+        elif tag is _Tag.DIRECT:
+            assert len(key) == 1 + 32 + 32
+            tx_hash = key[1:33]
+            tx_dep_hash = key[33:]
+            assert len(tx_hash) == len(tx_dep_hash) == 32
+            return _KeyAny(tag, tx_hash, tx_dep_hash)
+        elif tag is _Tag.REVERSE:
+            assert len(key) == 1 + 32 + 32
+            tx_dep_hash = key[1:33]
+            tx_hash = key[33:]
+            assert len(tx_hash) == len(tx_dep_hash) == 32
+            return _KeyAny(tag, tx_hash, tx_dep_hash)
+        elif tag is _Tag.READY:
+            assert len(key) == 1 + 32
+            tx_dep_hash = key[1:]
+            assert len(tx_dep_hash) == 32
+            return _KeyAny(tag, None, tx_dep_hash)
+        else:
+            # XXX: if/elif is exhaustive for all possible tags and invalid tag value will fail sooner
+            raise NotImplementedError('unreachable')
+
+    def _to_value_direct(self, validation: ValidationState) -> bytes:
+        import struct
+        value = struct.pack('!b', validation.value)
+        assert len(value) == 1
+        return value
+
+    def _from_value_direct(self, value: bytes) -> ValidationState:
+        import struct
+        assert len(value) == 1
+        validation_value, = struct.unpack('!b', value)
+        return ValidationState(validation_value)
+
+    def _to_value_needed(self, height: int, tx_hash: bytes) -> bytes:
+        import struct
+        value = bytearray()
+        value.extend(struct.pack('!I', height))
+        value.extend(tx_hash)
+        assert len(value) == 4 + 32
+        return bytes(value)
+
+    def _from_value_needed(self, value: bytes) -> Tuple[int, bytes]:
+        import struct
+        assert len(value) == 4 + 32
+        height, = struct.unpack('!I', value[:4])
+        tx_hash = value[4:]
+        return height, tx_hash
+
+    def update_validation(self, tx: bytes, validation: ValidationState) -> None:
+        import rocksdb
+        batch = rocksdb.WriteBatch()
+        if any(self._iter_rev_deps_of(tx)):
+            self._update_validation(tx, validation, batch)
+        self._db.write(batch)
+
+    def _update_validation(self, tx: bytes, validation: ValidationState, batch: 'rocksdb.WriteBatch') -> None:
+        """Propagate the new validation state of a given dep."""
+        new_value = self._to_value_direct(validation)
+        cousins = list(self._iter_rev_deps_of(tx))
+        for cousin in cousins:
+            key = self._to_key_direct(cousin, tx)
+            value = self._db.get((self._cf, key))
+            if value != new_value:
+                batch.put((self._cf, key), new_value)
+        # XXX: this only makes sense if the tx validation is complete
+        if validation.is_fully_connected():
+            for cousin in cousins:
+                if all(v.is_fully_connected() for _, v in self._iter_deps_of(cousin)):
```

Comment:
Because `validation` will most often not be "fully connected", and the if condition `all(v.is_fully_connected() for _, v in self._iter_deps_of(cousin))` is not very cheap. So I think it only makes sense to make those checks when `validation` is complete (aka "fully connected"). I think my `XXX` comment might have been too brief. Do you have a suggestion of a better comment?

Reply:
Now I get it. Maybe you can add just like you've written here

> Because validation will most often not be "fully connected", and the if condition all(v.is_fully_connected() for _, v in self._iter_deps_of(cousin)) is not very cheap. So I think it only makes sense to make those checks when validation is complete (aka "fully connected").

Reply:
Done.